### PR TITLE
Engine pr

### PR DESCRIPTION
**Pull request by BenjiPestell**

### CODE REVIEW

**Review by LettyYeti — CHANGES_REQUESTED** [review]
See doc, address clarifications & concerns prior to merge :)

**Review by LettyYeti — CHANGES_REQUESTED** [review]
See comment on coordinates for approval, other comment is ticketed but non-blocking

File: src/asmcnc/apps/drywall_cutter_app/engine.py [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+                raise Exception(u"Sir, rectangles have 4 sides, not %d" % len(rect_coordinates))
+
+            # Add first point to end of coordinate list to complete the contour
+            coordinates = rect_coordinates
+            coordinates.append(coordinates[0])
+
+            # Create a list of stepovers to add finishing passes
+            finish_passes = 1
+            finish_stepover = 0.5
+            finish_stepdown = self.config.active_config.cutting_depths.material_thickness
```
[the remark this answers is not in the document]
Comment (BenjiPestell) [reply src/asmcnc/apps/drywall_cutter_app/engine.py:484]: finish_stepdown = total_cut_depth

**Review by Masc999 — COMMENTED** [review]
Changes look good to me.

File: src/asmcnc/apps/drywall_cutter_app/engine.py [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+        if self.config.active_config.shape_type in file_structure_1_shapes:
+            output = "(%s)\nG90\nM3 S%d\nG0 %s\n\n%s\n(End)\nG0 Z%d\nM5\n" % (
+                output_file[output_file.find("/")+1:], self.config.active_cutter.cutting_spindle_speed, safe_start_position, ''.join(cutting_lines), z_safe_distance)
+        else:
+            output = ''.join(cutting_lines)  # Use ''.join() to concatenate lines without spaces
+
+        with open(output_file, 'w+') as out_file:
+            out_file.write(output.decode('utf-8'))  # Use write() to write the entire output as a single string
+            print("%s written" % output_file)
+
```
Comment (archiepeterjarvis) [inline src/asmcnc/apps/drywall_cutter_app/engine.py:597]: Some Logger.Debug through here could be nice :)

File: src/asmcnc/apps/drywall_cutter_app/test_engine.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        # Case 5: No adjustment needed for Z axis
+        gcode_lines = ["G1 X10 Y20", "G1 Z5", "G1 X30 Y40"]
+        x_adjustment = 5
+        y_adjustment = 10
+        expected_output = ["G1 X15 Y30", "G1 Z5", "G1 X35 Y50"]
+        output = self.engine.apply_datum_offset(gcode_lines, x_adjustment, y_adjustment)
+        self.assertEqual(output, expected_output)
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (archiepeterjarvis) [inline src/asmcnc/apps/drywall_cutter_app/test_engine.py:383]: Nice to see the test coverage 👍

**Pull request merged**